server/src/main/java/io/druid/server/BrokerQueryResource.java

```diff
@@ -102,4 +118,38 @@ public Response getQueryTargets(
       return context.gotError(e);
     }
   }
+
+  @Override
+  protected Query prepareQuery(Query query)
+  {
+    return rewriteDataSources(query);
+  }
+
+  private Query rewriteDataSources(Query query)
+  {
+    DataSource dataSource = query.getDataSource();
+    if (dataSource instanceof RegexDataSource) {
+      Set<String> found = Sets.newLinkedHashSet();
+      Iterable<DruidServer> inventory = serverInventoryView.getInventory();
+      for (String pattern : dataSource.getNames()) {
```

Comment:
Would it be more efficient to iterate through the server/datasources in the outer loops and check the regex matches inside, skipping further regex checks for a datasource if one matches already?